state/etcdstatedriver.go

```diff
@@ -105,46 +103,49 @@ func (d *EtcdStateDriver) Read(key string) ([]byte, error) {
 	ctx, cancel := context.WithTimeout(context.Background(), ctxTimeout)
 	defer cancel()
 
-	resp, err := d.KeysAPI.Get(ctx, key, &client.GetOptions{Quorum: true})
-	if err != nil || resp == nil || resp.Node == nil {
-		// Retry few times if cluster is unavailable
-		if err.Error() == client.ErrClusterUnavailable.Error() {
-			for i := 0; i < maxEtcdRetries; i++ {
-				resp, err = d.KeysAPI.Get(ctx, key, &client.GetOptions{Quorum: true})
-				if err == nil && resp.Node != nil {
-					break
-				}
-
-				// Retry after a delay
-				time.Sleep(time.Second)
-			}
-		} else {
-			return []byte{}, err
+	var err error
+	var resp *client.Response
+
+	for i := 0; i < maxEtcdRetries; i++ {
+		resp, err = d.KeysAPI.Get(ctx, key, &client.GetOptions{Quorum: true})
+		if err == nil {
+			return []byte(resp.Node.Value), nil
+		} else if client.IsKeyNotFound(err) {
+			return []byte{}, core.Errorf("Key not found")
+		} else if err.Error() == client.ErrClusterUnavailable.Error() {
+			// Retry after a delay
+			time.Sleep(time.Second)
```

Comment:
@yuva29 Any reason why you did not retain the original logic? Are you trying to anything beyond just placing the whole logic inside a single for loop?